packages/zoltan2/core/src/algorithms/color/Zoltan2_AlgHybrid2GL.hpp

```diff
@@ -0,0 +1,1532 @@
+    //
+    //  boundary_size will hold the size of the list of boundary vertices
+    //
+    //  verts_to_send_atomic will hold the list of vertices to send
+    //
+    //  verts_to_send_size_atomic will hold the size of the list of
+    //                            vertices to send.
+    //
+    virtual void constructBoundary(const size_t n_local,
+		    		   Kokkos::View<offset_t*, device_type> dist_offsets_dev,
+		                   Kokkos::View<lno_t*, device_type> dist_adjs_dev,
+				   typename Kokkos::View<offset_t*, device_type>::HostMirror dist_offsets_host,
+				   typename Kokkos::View<lno_t*, device_type>::HostMirror dist_adjs_host,
+                                   Kokkos::View<lno_t*, device_type>& boundary_verts,
+				   gno_t& boundary_size,
+				   Kokkos::View<lno_t*, 
+				                device_type, 
+						Kokkos::MemoryTraits<Kokkos::Atomic> > verts_to_send_atomic,
+				   Kokkos::View<size_t*, 
+				                device_type, 
+						Kokkos::MemoryTraits<Kokkos::Atomic>> verts_to_send_size_atomic) = 0;
+				   
+  protected:
+    RCP<const base_adapter_t> adapter;
+    RCP<GraphModel<base_adapter_t> > model;
+    RCP<Teuchos::ParameterList> pl;
+    RCP<Environment> env;
+    RCP<const Teuchos::Comm<int> > comm;
+    bool verbose;
+
+    
+  private:
+    //This function constructs the second ghost layer.
+    //It proceeds in rounds in order to keep messages
+    //under MPI's maximum send size.
+    //
+    //OUTPUT ARGS
+    //  ownedPlusGhosts holds the initial list of GIDs for all owned vertices.
+    //                  This function changes the order of the ghost GIDs in
+    //                  order to more easily construct the second ghost layer.
+    //
+    //  adjs_2GL holds the second ghost layer CSR adjacencies
+    //
+    //  offsets_2GL holds CSR offsets into the second ghost layer adjacencies.
+    //
+    //INPUT ARGS
+    //  owners holds the owning proc for a given vertex, indexed by local ID.
+    //
+    //  adjs is the CSR adjacencies of the local graph with a single ghost layer.
+    //
+    //  offsets is the CSR offsets of the local graph with a single ghost layer.
+    //
+    //  mapOwned translates from Owned GID to LID. We only need this translation
+    //           for owned vertices.
+    //
+    void constructSecondGhostLayer(std::vector<gno_t>& ownedPlusGhosts,
+                                   const std::vector<int>& owners,
+                                   ArrayView<const gno_t> adjs,
+                                   ArrayView<const offset_t> offsets,
+                                   RCP<const map_t> mapOwned,
+                                   std::vector< gno_t>& adjs_2GL,
+                                   std::vector< offset_t>& offsets_2GL) {
+      std::vector<int> sendcounts(comm->getSize(),0);
+      std::vector<gno_t> sdispls(comm->getSize()+1,0);
+      //loop through owners, count how many vertices we'll send to each processor
```

Comment:
What is the criterion for sending a vertex to another processor?  Why are we sending the vertices?